providers/ldap/ad_provider.go

```diff
@@ -0,0 +1,255 @@
+	ldapConfig.GroupDNField = providerSettings[GroupDnFieldSetting]
+	ldapConfig.GroupMemberUserAttribute = providerSettings[GroupMemberUserAttributeSetting]
+
+	authConfig.LdapConfig = ldapConfig
+}
+
+//GetProviderSettingList returns the provider specific db setting list
+func (l *LdapProvider) GetProviderSettingList(listOnly bool) []string {
+	var settings []string
+
+	settings = append(settings, AccessModeSetting)
+	settings = append(settings, DomainSetting)
+	settings = append(settings, GroupSearchDomainSetting)
+	settings = append(settings, LoginDomainSetting)
+	settings = append(settings, PortSetting)
+	settings = append(settings, UserSearchFieldSetting)
+	settings = append(settings, ServiceAccountUsernameSetting)
+	settings = append(settings, GroupSearchFieldSetting)
+	settings = append(settings, UserObjectClassSetting)
+	settings = append(settings, UserNameFieldSetting)
+	settings = append(settings, GroupObjectClassSetting)
+	settings = append(settings, UserLoginFieldSetting)
+	settings = append(settings, UserDisabledBitMaskSetting)
+	settings = append(settings, ServerSetting)
+	settings = append(settings, ServiceAccountPasswordSetting)
+	settings = append(settings, UserEnabledAttributeSetting)
+	settings = append(settings, GroupNameFieldSetting)
+	settings = append(settings, TlsSetting)
+	settings = append(settings, TimeoutSetting)
+	settings = append(settings, AllowedIdentitiesSetting)
+	settings = append(settings, GroupDnFieldSetting)
+	settings = append(settings, GroupMemberUserAttributeSetting)
+
+	return settings
+}
+
+//GetLegacySettings returns the provider specific legacy db settings
+func (l *LdapProvider) GetLegacySettings() map[string]string {
+	settings := make(map[string]string)
+	settings["accessModeSetting"] = AccessModeSetting
+	settings["allowedIdentitiesSetting"] = AllowedIdentitiesSetting
+	return settings
+}
+
+//GetConfig returns the provider config
+func (l *LdapProvider) GetConfig() model.AuthConfig {
+	log.Debug("In LDAP getConfig")
+
+	authConfig := model.AuthConfig{Resource: client.Resource{
+		Type: "config",
+	}}
+
+	authConfig.Provider = Config
+	authConfig.LdapConfig = *l.ldapClient.config
+
+	authConfig.LdapConfig.Resource = client.Resource{
+		Type: "ldapconfig",
+	}
+
+	log.Debug("In LDAP authConfig %v", authConfig)
+	return authConfig
+}
+
+//GetIdentitySeparator returns the provider specific separator to use to separate allowedIdentities
+func (l *LdapProvider) GetIdentitySeparator() string {
+	return "#"
+}
+
+// LoadConfig initializes the provider with the passed config
+func (l *LdapProvider) LoadConfig(authConfig *model.AuthConfig) error {
+	configObj := authConfig.LdapConfig
+	l.ldapClient.config = &configObj
+	return nil
+}
+
+//RefreshToken re-authenticates and generate a new token
+func (l *LdapProvider) RefreshToken(json map[string]string) (model.Token, error) {
+	log.Infof("%s IdentityProvider does not support RefreshToken API", Name)
```

Comment:
refreshToken() will get called from cattle: ExternalServiceAuthProvider, so we need to port the logic here from https://github.com/rancher/cattle/blob/91de2d9d20333501e582ac2a0ee8ed24eaac3046/code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/integration/ldap/ad/ADIdentityProvider.java#L61-L86

Test the auth using API key and Secret Key, and this code will be hit.